README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# flightplanner

<!-- badges: start -->

[![Project Status: WIP – Initial development is in progress, but there
has not yet been a stable, usable release suitable for the
public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)

<!-- badges: end -->

`flightplanner` is a tool for creating flight plans for use with the
Litchi software. It allows users to specify the latitude and longitude
of an origin point, as well as the dimensions, orientation, and altitude
of a survey area. The tool then calculates the best positions for
capturing photos of the survey area using a drone-mounted camera with a
vertical orientation. The resulting flight plan is designed to create an
orthomosaic from a series of overlapping images taken in a grid pattern.
`flightplanner` also provides visualization features for previewing the
flight plan and survey photo grid.

## Installation

You can install the development version of `flightplanner` like this:

``` r
# install package 'remotes' if necessary
# will already be installed if 'devtools' is installed
install.packages("remotes")
```

## Usage

First remember to load the package.

``` r
library(flightplanner)
```

To plan a mission, follow these steps:

1.  Determine the horizontal and vertical extent of your survey area,
    measured in meters from the origin point. For example, you might
    choose a horizontal limit of 100 meters to the east and a vertical
    extent of 200 meters to the north.
2.  Set the altitude of your survey area, measured in meters above the
    origin point. For example, you might choose an altitude of 60
    meters.
3.  Run the `fp_photo_grid` function with the `horiz`, `vert`, and
    `altitude` arguments from steps 1 and 2. Set the `plot` argument set
    to `TRUE` to visualize your photo survey grid in a plot. Modify the
    function’s settings (e.g. including desired overlap in vertical and
    horizontal directions) until you are satisfied with the layout of
    the grid.
4.  Decide on an origin point for your survey area, defined by its
    latitude and longitude coordinates (e.g. 55.125505, 10.268467). This
    will be the lower left point of your image grid.
5.  Decide on an angle by which the survey grid needs to be
    pivoted/rotated around this origin point (e.g. 45 degrees).
6.  Use the `fp_litchi_mission` function to generate a flight mission
    plan for use with the Litchi Mission Hub app. This mission plan is a
    `data.frame` which you save as a CSV file for use in the Litchi
    Mission Hub app.
7.  Check your plan on the Litchi MissionHub and adjust as necessary.

Be aware that the plan assumes that the topology is flat, that there are
no trees or other structures. Therefore, pay particular attention to the
altitude in areas that violate these assumptions.

Here’s a worked example.

First make a photo grid:

``` r
fp_photo_grid(altitude = 60, overlap_width = 0.6,
overlap_height = 0.4, survey_xaxis = 140, survey_yaxis = 300, plot = TRUE)
```

![](man/figures/README-photoGrid-1.png)<!-- -->

Then, when satisfied with this, run it again with `plot = FALSE` to get
the photo grid.

``` r
pg <- fp_photo_grid(altitude = 60, overlap_width = 0.6,
overlap_height = 0.4, survey_xaxis = 140, survey_yaxis = 300, plot = FALSE)
```

Next, produce the litchi plan:

``` r
mission1 <- fp_litchi_mission(pg, origin_lat = 55.125505, origin_long = 10.268467, angle = 38)
head(mission1)
#>        lat      lng altitude heading curve rotationdir gimbalmode gimbalangle
#> 1 55.12511 10.26934       60      38     0           0          2         -90
#> 2 55.12524 10.26905       60      38     0           0          2         -90
#> 3 55.12530 10.26960       60      38     0           0          2         -90
#> 4 55.12537 10.26876       60      38     0           0          2         -90
#> 5 55.12543 10.26931       60      38     0           0          2         -90
#> 6 55.12549 10.26985       60      38     0           0          2         -90
#>   actiontype1 actionparam1 actiontype2 actionparam2 altitudemode rowNumber
#> 1           0         1500           1            0            0         1
#> 2           0         1500           1            0            0         2
#> 3           0         1500           1            0            0         3
#> 4           0         1500           1            0            0         4
#> 5           0         1500           1            0            0         5
#> 6           0         1500           1            0            0         6
```

This file can then be exported like this, ready to be uploaded to the
Litchi Mission Hub.

    write.csv(x = mission1, file = "missionFiles/mission1.csv",
    row.names = FALSE)

## Note

WARNING: This package was developed for my own use and may contain bugs
or errors. Please use caution when using it, and carefully check your
plan using Litchi or Google Earth. The plans generated by this package
are ignorant of landscape topology, trees, buildings and other
structures, so please plan accordingly. The use of this package and the
resulting plans are your responsibility. I hope that it can be useful,
but use it at your own risk.

## Contributions

All contributions are welcome. Please note that this project is released
with a [Contributor Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By participating in this project you agree to abide by its terms.

There are numerous ways of contributing.

1.  You can submit bug reports, suggestions etc. by [opening an
    issue](https://github.com/jonesor/flightplanner/issues).

2.  You can copy or fork the repository, make your own code edits and
    then send us a pull request. [Here’s how to do
    that](https://jarv.is/notes/how-to-pull-request-fork-github/).

3.  You are also welcome to email me.


NEWS.md
# flightplanner 0.1.0

* The first version. It works.


R/drone_photoGrid.R
#' Create a grid of points representing the centers of photos
#'
#' @param altitude The altitude at which the photos will be taken (in meters).
#' @param overlap_width The width-wise overlap between photos (as a fraction between 0 and 1).
#' @param overlap_height The height-wise overlap between photos (as a fraction between 0 and 1).
#' @param survey_xaxis The width of the survey area (in meters).
#' @param survey_yaxis The height of the survey area (in meters).
#' @param plot A boolean value indicating whether to plot the photo grid (TRUE) or return the data frame of points (FALSE).
#' @param fov numeric, optional. Field of view of the camera in degrees. Default is 94.
#' @param aspectRatio numeric vector, optional. Aspect ratio of the image in the form of a 2-element vector, with the width as the first element and the height as the second element. Default is c(4,3).
#' @param ... additional arguments to be passed to the function.
#' @return If `plot` is TRUE, returns a `ggplot` object visualizing the photo grid. If `plot` is FALSE, returns a data frame of consisting of photoID and points representing the centers of each photo, and altitude.
#'
#' @importFrom dplyr mutate
#' @importFrom dplyr select
#' @importFrom dplyr desc
#' @importFrom tidyr expand_grid
#' @import ggplot2
#'
#' @export
#' @examples
#' fp_photo_grid()
#' fp_photo_grid(
#'   altitude = 60, overlap_width = 0.6,
#'   overlap_height = 0.2, survey_xaxis = 400, survey_yaxis = 500, plot = TRUE
#' )
#'
fp_photo_grid <- function(altitude = 40,
                          overlap_width = 0.4,
                          overlap_height = 0.1,
                          survey_xaxis = 200,
                          survey_yaxis = 300,
                          plot = TRUE,
                          fov = 94,
                          aspectRatio = c(4, 3),
                          ...) {

  # Validate inputs
  # Check that altitude is a positive number
  if (altitude <= 0) {
    stop("Altitude must be a positive number")
  }

  # Check that overlap width and overlap height are between 0 and 1
  if (overlap_width < 0 || overlap_width > 1) {
    stop("Overlap width must be between 0 and 1")
  }

  if (overlap_height < 0 || overlap_height > 1) {
    stop("Overlap height must be between 0 and 1")
  }

  # Check that plot is a boolean value
  if (!is.logical(plot)) {
    stop("Plot must be a boolean value (TRUE or FALSE)")
  }

  # Calculate the width of the footprint of each photo
  photo_footprint_width <- footprint_Width(altitude, fov = fov, aspectRatio = aspectRatio)

  # Calculate the distance between photos in the width-wise direction
  distanceBetweenPhotos_widthways <- photo_footprint_width - photo_footprint_width * overlap_width

  # Calculate the height of the footprint of each photo
  photo_footprint_height <- footprint_Height(altitude, fov = fov, aspectRatio = aspectRatio)

  # Calculate the distance between photos in the height-wise direction
  distanceBetweenPhotos_heightways <- photo_footprint_height - photo_footprint_height * overlap_height

  # Create a grid of points representing the centers of each photo
  # First line horizontal
  horiz <- seq(0, survey_xaxis, distanceBetweenPhotos_widthways)

  # First line vertical
  vert <- seq(0, survey_yaxis, distanceBetweenPhotos_heightways)

  # Expand the grid to create a matrix of points
  # Add the min/max edges for photo footprints
  photo_grid <- expand_grid(horiz, vert) |>
    mutate(xmin = horiz - (photo_footprint_width / 2), xmax = horiz + (photo_footprint_width / 2)) |>
    mutate(ymin = vert - (photo_footprint_height / 2), ymax = vert + (photo_footprint_height / 2)) |>
    mutate(photoID = as.factor(1:dplyr::n())) |>
    mutate(altitude = altitude)


  # If plot is set to TRUE, create a visual representation of the grid with ggplot
  if (plot == TRUE) {
    # Visualise the image grid, withe the photo overlaps
    (p <- ggplot(photo_grid, aes(x = horiz, y = vert)) +
      geom_point() +
      geom_rect(mapping = aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax, fill = photoID), alpha = 0.1) +
      coord_fixed() +
      ggtitle("Photo grid", subtitle = paste0("Altitude: ", altitude, "m; Gimball at -90 degrees")) +
      theme_minimal() +
      theme(legend.position = "none"))
    return(p)
  }
  # If plot is set to FALSE, return the data frame of points
  if (plot == FALSE) {
    return(photo_grid |>
      select(photoID, horiz, vert, altitude))
  }
}


R/drone_footprint_calculations.R
#' @title Drone Footprint Calculations
#'
#' @description Calculates various aspects of a drone's camera footprint at a given altitude, field of view, and image aspect ratio.
#'
#' @name footprint_
#' @param altitude numeric, required. Altitude of the drone in meters.
#' @param fov numeric, optional. Field of view of the camera in degrees. Default is 94.
#' @param aspectRatio numeric vector, optional. Aspect ratio of the image in the form of a 2-element vector, with the width as the first element and the height as the second element. Default is c(4,3).
#' @param ... additional arguments to be passed to the function.
#'
#' @return A single numeric value representing the calculated aspect of the camera footprint.
#'
#' @examples
#' \dontrun{
#' # Calculate the width of the camera footprint at an altitude of 100 meters
#' footprint_Width(altitude = 100)
#'
#' # Calculate the height of the camera footprint at an altitude of 50 meters
#' # and a field of view of 120 degrees
#' footprint_Height(altitude = 50, fov = 120)
#'
#' # Calculate the diagonal length of the camera footprint at an altitude of
#' # 200 meters, a field of view of 90 degrees, and an image aspect ratio of
#' # 4:3
#' footprint_Diagonal(altitude = 200, fov = 90, aspectRatio = c(4, 3))
#' }
#' @export

#' @rdname footprint_
footprint_Diagonal <- function(altitude, fov = 94, aspectRatio = c(4, 3), ...) {
  imageWidth <- aspectRatio[1]
  imageHeight <- aspectRatio[2]

  # Check that altitude is a positive number
  if (altitude <= 0) {
    stop("Altitude must be a positive number")
  }

  # Check that fov is between 0 and 180 degrees
  if (fov < 0 || fov > 180) {
    stop("Field of view must be between 0 and 180 degrees")
  }

  # Check that image dimensions are positive numbers
  if (imageHeight <= 0 || imageWidth <= 0) {
    stop("Image aspect ratio dimensions must be positive numbers")
  }

  fov_radians <- fov * pi / 180
  return(2 * tan(fov_radians / 2) * altitude)
}

#' @rdname footprint_
footprint_Height <- function(altitude, fov = 94, aspectRatio = c(4, 3), ...) {
  imageWidth <- aspectRatio[1]
  imageHeight <- aspectRatio[2]

  # Check that altitude is a positive number
  if (altitude <= 0) {
    stop("Altitude must be a positive number")
  }

  # Check that fov is between 0 and 180 degrees
  if (fov < 0 || fov > 180) {
    stop("Field of view must be between 0 and 180 degrees")
  }

  # Check that image dimensions are positive numbers
  if (imageHeight <= 0 || imageWidth <= 0) {
    stop("Image dimensions must be positive numbers")
  }

  fov_radians <- fov * pi / 180
  aspectRatio_fraction <- imageHeight / imageWidth
  d <- 2 * tan(fov_radians / 2) * altitude
  return(aspectRatio_fraction * d / sqrt(aspectRatio_fraction^2 + 1))
}

#' @rdname footprint_
footprint_Width <- function(altitude, fov = 94, aspectRatio = c(4, 3), ...) {
  imageWidth <- aspectRatio[1]
  imageHeight <- aspectRatio[2]

  # Check that altitude is a positive number
  if (altitude <= 0) {
    stop("Altitude must be a positive number")
  }

  # Check that fov is between 0 and 180 degrees
  if (fov < 0 || fov > 180) {
    stop("Field of view must be between 0 and 180 degrees")
  }

  # Check that image dimensions are positive numbers
  if (imageHeight <= 0 || imageWidth <= 0) {
    stop("Image dimensions must be positive numbers")
  }

  fov_radians <- fov * pi / 180
  aspectRatio_fraction <- imageHeight / imageWidth
  d <- 2 * tan(fov_radians / 2) * altitude
  return(d / sqrt(aspectRatio_fraction^2 + 1))
}


R/drone_createLitchiMissionGrid.R
#' Compute a rotation matrix for a given angle
#'
#' @param angle_rad The angle in radians for which to compute the rotation matrix.
#' @return A rotation matrix.
#' @keywords internal
#' @export
#'
computeRotationMatrix <- function(angle_rad) {
  matrix(c(cos(angle_rad), -sin(angle_rad), sin(angle_rad), cos(angle_rad)), nrow = 2)
}

#' Convert a data frame to an sf object with WGS 84 coordinate reference system (CRS)
#'
#' @param data A data frame containing latitude and longitude columns.
#' @return An sf object with WGS 84 CRS.
#' @keywords internal
#'
#' @importFrom sf st_as_sf
#'
#' @export
#' @examples
#' convertToSf(data.frame(latitude = 45.5236, longitude = -122.675))
#'
convertToSf <- function(data) {
  st_as_sf(data, coords = c("longitude", "latitude"), crs = 4326, agr = "constant")
}

#' Transform an sf object to a meter-based CRS
#'
#' @param data An sf object.
#' @return The input sf object transformed to a meter-based coordinate reference system (CRS).
#' @keywords internal
#'
#' @importFrom sf st_transform
#'
#' @export
#' @examples
#' data <- data.frame(latitude = 45.5236, longitude = -122.675)
#' data <- convertToSf(data)
#' transformSf(data)
transformSf <- function(data) {
  st_transform(data, 3857)
}

#' Generate a mission plan for a drone in Litchi app
#'
#' @param photo_grid A photo grid data frame created using the `fp_photo_grid` function.
#' @param origin_lat The latitude of the starting point for the mission (in degrees).
#' @param origin_long The longitude of the starting point for the mission (in degrees).
#' @param angleDeg The angle at which the survey grid is rotated from North (in degrees). e.g. a value of 45 rotates the photo grid around the origin point (bottom-left) by 45 degrees.
#' @param ... additional arguments to be passed to the function.
#' @return A data frame containing the mission plan for the drone, in a format suitable for use with the Litchi app.
#'
#' @importFrom sf st_as_sf
#' @importFrom sf st_coordinates
#' @importFrom dplyr mutate
#' @importFrom dplyr select
#' @importFrom dplyr filter
#' @importFrom dplyr arrange
#' @importFrom dplyr bind_rows
#'
#' @export
#' @examples
#' photo_grid <- fp_photo_grid(
#'   altitude = 60, overlap_width = 0.6,
#'   overlap_height = 0.2, survey_xaxis = 100, survey_yaxis = 200, plot = FALSE
#' )
#'
#' litchiMission <- fp_litchi_mission(photo_grid, pg,
#'   origin_lat = 55.125505,
#'   origin_long = 10.268467, angle = 38
#' )
#' \dontrun{
#' write.csv(
#'   x = litchiMission, file = "missionFiles/testGrid10.csv",
#'   row.names = FALSE
#' )
#' }
#'
fp_litchi_mission <- function(photo_grid, origin_lat, origin_long, angleDeg = 0, ...) {

  # Validation of inputs
  # photo_grid, should be a data frame containing columns for photoID, horiz, vert, and altitude
  if (!inherits(photo_grid, "data.frame")) {
    stop("photo_grid must be a data frame")
  }
  if (!is.numeric(photo_grid$horiz) || !is.numeric(photo_grid$vert)) {
    stop("photoID, horiz, and vert must be numeric columns in photo_grid")
  }
  if (length(unique(photo_grid$photoID)) != nrow(photo_grid)) {
    stop("photoID must be a unique identifier for each row in photo_grid")
  }
  if (min(photo_grid$altitude) <= 0) {
    stop("altitude must contain a positive value in photo_grid")
  }
  if (any(photo_grid$horiz < 0) || any(photo_grid$vert < 0)) {
    stop("horiz and vert cannot be negative in photo_grid")
  }


  # Check that angle is between 0 and 360 degrees
  if (angleDeg < 0 || angleDeg > 360) {
    stop("Angle must be between 0 and 360 degrees")
  }

  # Set projection of origin point as coordinate reference system (CRS) WGS 84 (crs 4326)
  origin_sf <- convertToSf(data.frame(latitude = origin_lat, longitude = origin_long))

  # Convert origin point to meter-based reference system
  origin1_sf_m <- transformSf(origin_sf)

  # Convert angle from degrees to radians
  angle_rad <- angleDeg * pi / 180

  # Compute the rotation matrix
  R <- computeRotationMatrix(angle_rad)

  # Create a photo grid, based on overlap, height and survey extent
  photo_grid_points <- photo_grid |> select(horiz, vert)

  photo_grid_altitude <- photo_grid$altitude

  # Rotate each point in the data frame
  rotated_grid_points <- as.data.frame(t(apply(photo_grid_points, 1, function(p) R %*% as.matrix(p))))
  names(rotated_grid_points) <- c("x", "y")

  photo_grid_points$x_rot <- rotated_grid_points$x
  photo_grid_points$y_rot <- rotated_grid_points$y

  # Place the photo grid onto on lat long
  photo_grid_points <- photo_grid_points |>
    select(x_rot, y_rot) |>
    mutate(long = x_rot + st_coordinates(origin1_sf_m)[, "X"]) |>
    mutate(lat = y_rot + st_coordinates(origin1_sf_m)[, "Y"])

  photo_grid_points_sf <- st_as_sf(photo_grid_points,
    coords = c("long", "lat"),
    crs = 3857, agr = "constant"
  )

  photo_grid_points_sf_latlong <- st_transform(photo_grid_points_sf, 4326)

  # Add the X/Y coordinates to the dataframe
  photo_grid_points_sf_latlong$lng <- st_coordinates(photo_grid_points_sf_latlong)[, "X"]
  photo_grid_points_sf_latlong$lat <- st_coordinates(photo_grid_points_sf_latlong)[, "Y"]

  # Litchi file creation
  litchiMission <- data.frame(
    lat = photo_grid_points_sf_latlong$lat, lng = photo_grid_points_sf_latlong$lng,
    altitude = photo_grid_altitude,
    heading = angleDeg,
    curve = 0, rotationdir = 0, gimbalmode = 2, gimbalangle = -90,
    actiontype1 = 0, actionparam1 = 1500,
    actiontype2 = 1, actionparam2 = 0,
    altitudemode = 0
  ) |> mutate(rowNumber = as.numeric(as.factor(lat)))

  # Some rearranging, necessary to ensure an efficient "lawnmower" flight path.
  litchiMission_oddRowID <- seq(1, max(litchiMission$rowNumber), 2)

  # The odd rows...
  litchiMission_odd <- litchiMission |>
    filter(rowNumber %in% litchiMission_oddRowID)

  # The even rows (sorted in decending longitude order)...
  litchiMission_even <- litchiMission |>
    filter(!rowNumber %in% litchiMission_oddRowID) |>
    arrange(desc(lng))

  # Recombine the two parts of the data frame
  litchiMission <- bind_rows(litchiMission_odd, litchiMission_even) |>
    arrange(lat)

  rownames(litchiMission) <- NULL

  return(litchiMission)
}
